[risk=low][RW-9390] Added cromwell configuration panel.

evrii opened the pull request:
Description:
Added a configuration panel that users will see as the last step before creating Cromwell app.

https://user-images.githubusercontent.com/24514630/215195077-5872ec2e-d103-47e3-a7ef-4a830388e82b.mov

<!--
Replace this template with your PR description.
Please remember to keep in mind the security levels outlined in
[CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md) and to
include a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title

* **no**: None
* **low**: Low chance of potential impact to, or exposure of patient data
* **moderate**: Moderate chance of potential impact to, or exposure of patient data
* **severe**: Severe chance of potential impact to, or exposure of patient data

Please also:

* Get thumbs from reviewer(s)
* Verify all tests go green, including CI tests
-->


---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

evrii commented on ui/src/app/components/cost-predictor.tsx:
```diff
@@ -0,0 +1,85 @@
+  workspace,
+  analysisConfig,
+  status,
+  onPause,
```
Is there a better way of doing getting on Pause and onResume to StartStopRuntimeButton rather than passing them through CostPredictor? Could we use context or a utility function or something?

Peter-Lavigne replied:
I don't think React is the issue here. If `CostPredictor` does more than just predict costs, maybe the component name should change instead? Or perhaps the component is doing too much?

jmthibault79 replied:
can you pass the StartStopRuntimeButton as a child component?  This would also simplify things because you can keep StartStopRuntimeButton runtime-specific (no changes) and pass in a "FakeCromwellStartButton" as a placeholder

evrii replied:
Hmmm, I like the idea of separating them, because they are quite distinct in their functionality, but since the two subcomponents are only ever used from within this component (not on their own), I will go the route of renaming.

I am willing to go the other way, if anyone would prefer that approach.

evrii replied:
EnvironmentInformedActionPanel. A little wordy, but I think it summarizes what the intent of the component is.

evrii commented on ui/src/app/utils/runtime-utils.tsx:
```diff
@@ -111,13 +111,14 @@ export interface AnalysisConfig {
   machine: Machine;
   diskConfig: DiskConfig;
-  // This should only be populated if !diskconfig.detachable.
+  // This should only be populated if !diskconfig.detachable. (What does this mean?)
```
I updated the comment here, so I could ask about what this means.

Firstly, is a detachable disk a persistent disk?

DiskConfig has a property called detachable. I would imagine that this property designates whether a disk is detachable/persistent.

If that is the case, what is the detachedDisk mentioned below? The comment above says that it should only be set if diskConfig is not detachable.

evrii replied:
After talking about this as a team, I created a spike. I will update the comment with reference to the spike.

evrii commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,196 @@
+import { findMachineByName } from 'app/utils/machines';
+import { runtimeStore, useStore } from 'app/utils/stores';
+
+import { defaultCromwellConfig } from './apps-panel/utils';
```
Should this file be located somewhere else? I am wondering if it is worth having a configuration panel directory.

Peter-Lavigne replied:
I'm fine with either. It's easy to change this later as well.

jmthibault79 replied:
`components/runtime-configuration-panel/` and `components/apps-panel/`both exist btw.

evrii replied:
Are you suggesting that we should put it in one of those directories, or are you suggesting that there is convention for making directories?

jmthibault79 replied:
I have not thought about it enough to make a specific suggestion - and now I see that obviously you know about them because you're updating them here.  never mind :)

evrii commented on ui/src/app/components/help-sidebar.tsx:
```diff
@@ -628,6 +681,27 @@ export const HelpSidebar = fp.flow(
+                              alignItems: 'center',
+                            }}
+                          >
+                            <Button
```
Currently not an easy way of testing this functionality automatically in a unit test. The cromwell configuration panel test does not load the header and the help sidebar tests do not load the sidebar pabel.

evrii replied:
No longer relevant with the back button removed.

Peter-Lavigne commented on ui/src/app/components/configuration-panel.tsx:
```diff
@@ -0,0 +1,62 @@
+  );
+});
+
+export const ConfigurationPanel = ({
+  type = 'runtime',
+  onClose,
+  runtimeConfPanelInitialState,
+}) => {
+  return <PanelMain {...{ type, onClose, runtimeConfPanelInitialState }} />;
+};
```
Could we remove this in favor of exporting + renaming PanelMain?

Peter-Lavigne commented on ui/src/app/components/runtime-configuration-panel.tsx:
```diff
@@ -630,8 +559,12 @@ const PanelMain = fp.flow(
                 <CreatePanel
-                  creatorFreeCreditsRemaining={creatorFreeCreditsRemaining}
-                  profile={profile}
+                  {...{
+                    creatorFreeCreditsRemaining,
+                    profile,
+                    setRuntimeStatus,
+                    status,
+                  }}
```
Minor: Personally, I find the `...` notation mixed with non-`...` notation in the same props call confusing. I keep accidentally parsing it as a single object argument instead of several independent props. I think either approach (`...` or no `...`) is fine, but would prefer to not have them combined.

jmthibault79 replied:
and here you can set all the props that way :)

evrii replied:
Cleaned this up. Thanks for the feedback.

Peter-Lavigne commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,204 @@
+          <WarningMessage>
+            This cost is only for running the Cromwell Engine, there will be
+            additional cost for interactions with the workflow.
+            <a href='#' style={{ marginLeft: '0.25rem' }}>
```
These help links don't appear to go anywhere. Is that intentional?

evrii replied:
Yeah, that is intentional. I believe that the actual links are coming later.

jmthibault79 replied:
we should file a story to track that

Peter-Lavigne reviewed (changes requested):
Overall, I think this looks solid, especially sharing components between apps and notebooks.

I did notice a couple things:
- I can spam the start button to create several apps.
- The runtime status on the config page isn't updating based on cromwell's status, as far as I can tell.

Maybe my local environment isn't set up correctly or my workspaces are bugged; I haven't done any dev work with apps yet. I can hop on a call anytime today to show you.

Also, I haven't reviewed any tests yet.

evrii commented on ui/src/app/components/runtime-configuration-panel/start-stop-runtime-button.tsx:
```diff
@@ -1,31 +1,56 @@
+  onResume,
+  appType,
+}: StatusInfo) => {
+  // Get status in shared type
+  //  Update StartStopRunTimeButton to expect status and a way to update it
+  // Update instances of RuntimeStatus to shared type
+  // setup shared way of updating status - did Joel already make this?
+  // Make sure Cromwell always shows not sign for now
```
These were notes to myself. Thanks for catching.

jmthibault79 commented on ui/src/app/components/runtime-configuration-panel/start-stop-runtime-button.tsx:
```diff
@@ -46,18 +71,16 @@ export const StartStopRuntimeButton = ({
     ],
     [
-      RuntimeStatus.Running,
+      'Running',
```
That hasn't been my experience - I have seen TS enforce spelling mistakes here.  Do you have an example?

Peter-Lavigne replied:
When I switch `'Running'` to `'not-a-status'` the app compiles.

The main reason for this is that `switchCase` is untyped. However, when I add a type to `switchCase` such as `<T, U>(value: T, ...pairs: [T, () => U][]): U`, TS infers T to be `UserEnvironmentStatus` when the matched value is `'Running'` but infers T to be a string when the matched value is `'not-a-status'`.

jmthibault79 replied:
I see, thanks.  I had been favoring string unions recently over enums, but now I'm not so sure.

`UserEnvironmentStatus` is quite new, so it shouldn't be terribly intrusive to switch to an enum now.  I'd be fine with that.

evrii replied:
I changed UserEnvironmentStatus from a union to an enum.

jmthibault79 commented on ui/src/app/components/runtime-configuration-panel/start-stop-runtime-button.tsx:
```diff
@@ -1,31 +1,56 @@
+  appType: UIAppType;
+}
+
 export const StartStopRuntimeButton = ({
```
Yes, we should not refer to runtimes specifically, if this is now generic.  We should also update other variables, text and ID, like:
```
altText: 'Runtime running, click to pause',
        iconSrc: computeRunning,
        dataTestId: 'runtime-status-icon-running',
        onClick: onPause,
```
```
altText: 'Environment running, click to pause',
        iconSrc: computeRunning,
        dataTestId: 'environment-status-icon-running',
        onClick: onPause,
```
etc.

But also: I wonder if all this work is needed at this time.  Because this button won't be operational for Cromwell MVP, right?  Maybe we should revert these changes and save the work for the real Cromwell Config Panel.

evrii replied:
Was not a hard refactor. Was able to do it decently quickly.

jmthibault79 commented on ui/src/app/components/runtime-configuration-panel.tsx:
```diff
@@ -82,66 +79,19 @@ import {
 
-const { useState, useEffect, Fragment } = React;
-
-const CostInfo = ({
```
I like the extraction of the cost elements into CostPredictor

evrii replied:
Thanks :-)

jmthibault79 commented on ui/src/app/components/runtime-configuration-panel.spec.tsx:
```diff
@@ -1109,32 +1109,32 @@ describe('RuntimeConfigurationPanel', () => {
     // Default GCE machine, n1-standard-4, makes the running cost 20 cents an hour and the storage cost less than 1 cent an hour.
-    expect(runningCost(wrapper).text()).toEqual('$0.20/hour');
-    expect(storageCost(wrapper).text()).toEqual('< $0.01/hour');
+    expect(runningCost(wrapper).text()).toEqual('$0.20per hour');
```
What do you think about adding a space before `per hour` in the component?  That would allow you to remove its marginLeft, and also its `text()` would look more correct.

(this is also fine)

jmthibault79 commented on ui/src/app/components/help-sidebar-icons.tsx:
```diff
@@ -350,6 +352,30 @@ const displayIcon = (icon: IconConfig, props: DisplayIconProps) => {
       () => displayRuntimeStatusIcon(icon, workspace.namespace),
     ],
+    [
+      'cromwellConfig',
```
this looks good

jmthibault79 commented on ui/src/app/components/help-sidebar-icons.tsx:
```diff
@@ -374,6 +400,7 @@ const displayIcon = (icon: IconConfig, props: DisplayIconProps) => {
       () =>
         icon.faIcon === null ? (
           <img
+            alt={icon.id}
```
I think label would be better here

jmthibault79 commented on ui/src/app/components/help-sidebar-icons.tsx:
```diff
@@ -490,6 +517,16 @@ const iconConfig = (props: IconConfigProps): IconConfig => {
+    cromwellConfig: {
+      id: 'cromwellConfig',
+      disabled: !!loadingError,
+      faIcon: faEdit,
```
`null` is better here, so as not to mislead anyone

jmthibault79 commented on ui/src/app/components/help-sidebar.tsx:
```diff
@@ -373,6 +381,7 @@ export const HelpSidebar = fp.flow(
       bodyPadding?: string;
       renderBody: () => JSX.Element;
       showFooter: boolean;
+      showNavigation?: boolean;
```
is this intended to do more than show the back button?  if not, maybe `showBackNavigation`?

I'm not sure that it's the right choice to have a back button, however.  We're assuming here that "back" means the Apps Panel, but I can also navigate to the Cromwell Config Panel by clicking the Cromwell icon.  For this flow, "back" is not meaningful, or it means "close"

evrii replied:
I don't have strong preferences about the back button. I was thinking that the navigation was something that we might be building out over time. I saw it in the mockup:
https://www.sketch.com/s/64a4f7f4-b8ca-41c7-91f9-b7d1de397a2b/a/WKJKGxw

If it doesn't make sense, I am good with taking it out.

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,204 @@
+  return (
+    <FlexColumn id='cromwell-configuration-panel' style={{ height: '100%' }}>
+      <div>
+        A cloud environment consists of application configuration, cloud compute
```
add "an" -> "of an application configuration"

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,204 @@
+    return (
+      <FlexColumn style={{ height: '100%' }}>
+        <div
+          data-test-id='runtime-create-panel'
```
cromwell-create-panel or similar

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,204 @@
+  creatorFreeCreditsRemaining = null,
+}) => {
+  const [analysisConfig, setAnalysisConfig] = useState({});
+  const { runtimeLoaded } = useStore(runtimeStore);
```
this is Jupyter - we don't need it here

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,153 @@
+  };
+
+  beforeEach(async () => {
+    runtimeApiStub = new RuntimeApiStub();
```
should be able to remove anything related to RuntimeApi and runtimeStore

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,153 @@
+  });
+
+  afterEach(() => {
+    act(() => clearCompoundRuntimeOperations());
```
shouldn't need this

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,153 @@
+  });
+
+  it('should show loading spinner while loading', async () => {
+    runtimeStore.set({
```
shouldn't need this

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,153 @@
+    expect(wrapper.exists(Spinner)).toEqual(false);
+  });
+
+  it('start button should create cromwell and close panel', async () => {
```
good test

jmthibault79 commented on ui/src/app/components/cost-predictor.tsx:
```diff
@@ -0,0 +1,85 @@
+    <StartStopRuntimeButton {...{ status, onPause, onResume, appType }} />
+    <CostInfo
+      {...{ analysisConfig }}
+      runtimeChanged={false}
```
this change caused a regression: previously, the Runtime Config Panel would change color when the user made a change.  Now it does not.

Should also be updated to `userEnvChanged` or something like it, to be more generic

jmthibault79 commented on ui/src/app/components/cost-predictor.tsx:
```diff
@@ -0,0 +1,85 @@
+            : {}),
+        }}
+      >
+        <RuntimeCostEstimator {...{ analysisConfig }} />
```
If this handles more than just runtimes now, we should probably rename it

evrii replied:
Renamed it to EnvironmentCostEstimator

jmthibault79 commented on ui/src/app/components/configuration-panel.tsx:
```diff
@@ -0,0 +1,54 @@
+import { RuntimeConfigurationPanel } from './runtime-configuration-panel';
+
+export const ConfigurationPanel = fp.flow(
+  withCdrVersions(),
```
is this used?

jmthibault79 commented on ui/src/app/components/apps-panel/utils.tsx:
```diff
@@ -94,20 +94,30 @@ export const findApp = (
 
 // used as a generic equivalence for certain states of RuntimeStatus and AppStatus
 export type UserEnvironmentStatus =
+  | 'Creating'
```
woo thanks!  Would you mind moving UNKNOWN to either the start or the end?

jmthibault79 reviewed (commented):
Tested locally, works well.  But I have a few comments and concerns.

The "cloud compute profile" is incorrect: it says 120 GB disk where our default is 50.  Could we derive these values from the defaultCromwellConfig?

When a user clicks Start, there's no indication that Cromwell has started, besides the panel closing.  Any ideas?

evrii commented:
> Overall, I think this looks solid, especially sharing components between apps and notebooks.
> 
> I did notice a couple things:
> 
> * I can spam the start button to create several apps.
> * The runtime status on the config page isn't updating based on cromwell's status, as far as I can tell.
> 
> Maybe my local environment isn't set up correctly or my workspaces are bugged; I haven't done any dev work with apps yet. I can hop on a call anytime today to show you.
> 
> Also, I haven't reviewed any tests yet.

Good points. I will look figure out how to address the start button issue.

As far as the runtime status goes, that is intentional for this ticket. This can be updated in the functional version of this page.

evrii commented:
> Tested locally, works well. But I have a few comments and concerns.
> 
> The "cloud compute profile" is incorrect: it says 120 GB disk where our default is 50. Could we derive these values from the defaultCromwellConfig?
> 
> When a user clicks Start, there's no indication that Cromwell has started, besides the panel closing. Any ideas?

I updated the profile to reflect the correct values. They were hard coded before.

I also updated the start button behavior. After you click it, it will close the cromwell panel. After a three second delay it will open the apps panel which should show you the current status. I would like to add a toast notification, but I do not think that will be in this ticket. I want to add a general functionality for us to send messages to a Toast from anywhere in the application.

Peter-Lavigne commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,122 @@
+  let onClose: () => void;
+  let freeTierBillingAccountId: string;
+
+  const component = async (propOverrides?: object) => {
+    const allProps = { ...props, ...propOverrides };
+    const c = mountWithRouter(<CromwellConfigurationPanel {...allProps} />);
+    await waitOneTickAndUpdate(c);
+    return c;
+  };
```
I really like "default props + overrides" functions in react tests like this 👍 

One minor suggestion would be to define a `DEFAULT_PROPS` constant to have only one way to define overridden props (currently you could define `propOverrides` or mutate `props`). I personally find it easier to understand at a glance than diving into `beforeEach`.

evrii commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -135,9 +145,12 @@ const PanelMain = fp.flow(
               onClose();
+              setTimeout(() => setSidebarActiveIconStore.next('apps'), 3000);
             }}
+            disabled={!!status || creationStarted}
```
Found a condition I hadn't considered. Working through this.

evrii replied:
Wasa ble to fix this/

jmthibault79 reviewed (changes requested):
Not sure what changed, but I can no longer click the Start button on the cromwell-config-panel

Peter-Lavigne commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,245 @@
+    onClose,
+  }) => {
+    // all apps besides Jupyter
+    const [userApps, setUserApps] = useState<UserAppEnvironment[]>();
+    const [creationStarted, setCreationStarted] = useState(false);
+    const [loading, setLoading] = useState(true);
+    const [status, setStatus] = useState<AppStatus>();
+
+    useEffect(() => {
+      appsApi().listAppsInWorkspace(workspace.namespace).then(setUserApps);
+    }, []);
+
+    useEffect(() => {
+      if (userApps !== undefined) {
+        const app = findApp(userApps, UIAppType.CROMWELL);
+        if (app) {
+          setStatus(app.status ?? null);
+        } else {
+          setLoading(false);
+        }
+      }
+    }, [userApps]);
+
+    useEffect(() => {
+      if (loading && status !== undefined) {
+        setLoading(false);
+      }
+    }, [loading, status]);
+
+    useEffect(() => {
+      if (creationStarted) {
+        appsApi().createApp(workspace.namespace, defaultCromwellConfig);
+        onClose();
+        setTimeout(() => setSidebarActiveIconStore.next('apps'), 3000);
+      }
+    }, [creationStarted]);
```
There is a lot of state and `useEffect` calls here. Could it be simplified? Maybe something like the following:

```
const [creationStarted, setCreationStarted] = useState(false);
const status = findApp(userApps, UIAppType.CROMWELL);
const loading = !app.status;

useEffect(() => {
  appsApi().listAppsInWorkspace(workspace.namespace).then(setUserApps);
}, []);
```

This should update the component every time `findApp` changes value.

Peter-Lavigne replied:
Additionally, it might be simpler to move the fourth `useEffect` into the function that calls `setCreationStarted(true)`.

evrii replied:
Good call. I like this refactor. Thanks for the suggestion.

evrii commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,122 @@
+      .find('#cromwell-cloud-environment-create-button')
+      .first();
+    startButton.simulate('click');
+    expect(spyCreateApp).toHaveBeenCalledTimes(1);
```
Followed up in chat.

Peter-Lavigne reviewed (approved):
Seems to be working fine locally for me.

For future large PRs, I've been experimenting with splitting them into multiple PRs to make review easier, if possible. 1000+ lines changed is a lot to review and can make the dev review step complicated and incomplete. As an example, the notebook / app component refactors might have been possible to move into another PR.

jmthibault79 commented on ui/src/app/components/runtime-configuration-panel/disabled-panel.tsx:
```diff
@@ -7,7 +7,7 @@ import { AoU } from 'app/components/text-wrappers';
   return (
     <WarningMessage
-      data-test-id='runtime-disabled-panel'
+      data-test-id='environment-disabled-panel'
```
please update the text below here: notebooks -> notebooks or user apps, or similar

jmthibault79 commented on ui/src/app/components/configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,76 @@
+    jest.useRealTimers();
+  });
+
+  it('should render disabled panel when creator billing disabled', async () => {
```
could you test the reverse as well?

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.spec.tsx:
```diff
@@ -0,0 +1,120 @@
+  });
+
+  afterEach(() => {
+    // jest.clearAllMocks();
```
can remove this line

jmthibault79 commented on ui/src/app/components/cromwell-configuration-panel.tsx:
```diff
@@ -0,0 +1,233 @@
+  }) => {
+    // all apps besides Jupyter
+    const [userApps, setUserApps] = useState<UserAppEnvironment[]>();
+    const [creationStarted, setCreationStarted] = useState(false);
```
nit: I like "creating" but this is fine too
